src/coreclr/jit/optimizer.cpp

```diff
@@ -1798,6 +1800,70 @@ bool Compiler::optTryUnrollLoop(FlowGraphNaturalLoop* loop, bool* changedIR)
     return true;
 }
 
+//-----------------------------------------------------------------------------
+// optReplaceScalarUsesWithConst: Replace all GT_LCL_VAR occurrences of a local
+// with a constant.
+//
+// Arguments:
+//   block   - The block to replace in
+//   lclNum  - The local to replace
+//   cnsVal  - The constant to replace with
+//
+// Remarks:
+//   This is used to replace the loop iterator with the constant value when
+//   unrolling.
+//
+void Compiler::optReplaceScalarUsesWithConst(BasicBlock* block, unsigned lclNum, ssize_t cnsVal)
```

Comment:
I am not familair with the existing logic - can you explain why it's fine to do propagation like this? can the locals be address exposed?

Reply:
`FlowGraphNaturalLoop::AnalyzeIteration` checks for this (it guarantees that the iterator local is only defined once within the loop and that the loop invariant is always true).